### ISSUE TITLE
Consolidiffication of DNS requests for a host.

### ISSUE THREAD

**Issue by Jorropo** [libp2p/go-libp2p#1034]
# TL;DR:
While implementing DNS over Tor for anonymity we faced the impossibilities to proxy the DNS requests made by libp2p and libs using it, we should try to fix that.

## Current status

I've identified 3 resolver spots and each one has a problem :
- [`basichost#Host.maResolver`](https://github.com/libp2p/go-libp2p/blob/fcf69647e945b30851fe394d140f7a70643e48cf/p2p/host/basic/basic_host.go#L92): no codepath available to set it with a custom `DialFunc` or resolver, underused.
- [`go-dnslink#defaultResolver`](https://github.com/ipfs/go-dnslink/blob/0626e50cfeaf0cd304eb0017de996e8cacb08540/dnslink.go#L115): no codepath available to set it, unsafe to set (racy), mixed Parser + Resolver object
- [`go-multiaddr-dns#DefaultResolver`](https://github.com/multiformats/go-multiaddr-dns/blob/aeb5743691b968cfa3365c9da59ef872a3133c87/resolve.go#L12): unsafe to set (racy)

---
### Important (non breaking, or minor breaking):

- go-dnslink:
	- split the dns / parsing part, sets the default exposed methods to use `go-multiaddr-dns#DefaultResolver` as backend.

### Need more discussion:

- Basic Host:
	- Add an `DNSResolver` options to `go-libp2p#New`, this would sets the `basichost#Host.maResolver` with a custom `*madns.Resolver` as argument.
	- Changes all internal usages in a host to using this internal resolver (might already be done idk, I will have to check).
- `host#Host`
	- Exposes an API on the interface (maybe an optional extension of the interface ?) for dns, madns and dnslink releated requests.
	- Change go-ipfs to use this new API.

---

When we came to a consensus about that I'll be happy to implement it (I'll probably throw an ugly api and many replace go.mod rules since).

**Comment by aschmahmann** [comment]
I agree that having a more configurable way to issue DNS requests that doesn't involve setting a global variable is a reasonable idea 👍.

However, I think a good deal of this is either done already or is an application level issue (e.g. ipfs/go-ipfs#6532).

- go-dnslink: I'm up for accepting changes here but AFAICT nobody is using is this library. Perhaps with some changes here we could use it instead of separately doing DNSLink resolution in go-ipfs's namesys module.
- BasicHost: It seems like this can be set by utilizing a custom MultiaddrResolver https://github.com/libp2p/go-libp2p/blob/fcf69647e945b30851fe394d140f7a70643e48cf/p2p/host/basic/basic_host.go#L230

@Jorropo is there something I'm missing here, and if not is this issue closable?

**Comment by Jorropo** [comment]
> BasicHost: It seems like this can be set by utilizing a custom MultiaddrResolver

There isn't any [`go-libp2p#Option`](https://pkg.go.dev/github.com/libp2p/go-libp2p#Option) who sets `opts.MultiaddrResolver`, I guess I can add one.

**Comment by Stebalien** [comment]
Fixed by https://github.com/libp2p/go-libp2p/blob/d38694b8cd1bf8598cc91258f8ca36ccd03a1fe7/options.go#L469.
